### PR TITLE
fix: log warning when image cannot be rendered (source key absent)

### LINKED ISSUE #233: Missing source key in image type objects
### Question
Hi there, I’m using this library and have a question regarding the output.

I noticed that for certain PDF pages containing images, the JSON output correctly identifies the bbox coordinates, but the image is not being saved/rendered in the Markdown output. Upon investigation, it seems this happens when the source key is missing from the image object.

I would like to know what causes the presence or absence of the source value for different images. Please note that I have already set the content_safety_off='all' option.

Image

Thank you for your help!

### PR DESCRIPTION
**Pull request by bundolee**
## Summary
Fixes #233

- Images detected in a PDF (with bbox) may not be extracted when the page renderer cannot produce a sub-image
- This caused `"source"` to be silently absent from JSON output with no feedback to the user
- Added `WARNING` log messages so users can identify which images failed and why

## Problem
`ImagesUtils.createImageFile()` called `ContrastRatioConsumer.getPageSubImage()` and silently returned when the result was `null` — no log, no error. The `source` key was then absent in JSON output for affected images, and the image did not appear in Markdown. The `--content-safety-off` option (which only controls text filtering) had no effect.

## Changes
- `ImagesUtils.java` — split the null check into two distinct warnings: `contrastRatioConsumer == null` (renderer initialization failure, affects all images in the file) vs `getPageSubImage() == null` (per-image render failure); use parameterized log calls throughout
- `ImagesUtilsTest.java` — test for the renderer-unavailable warning path (`testWriteImageLogsWarningWhenRendererIsUnavailable`); logger level is saved/restored; Files.walk() is closed in try-with-resources
- `json-schema.mdx` — document all conditions under which `source` may be absent, clarify embedded mode behavior (Base64 failure leaves neither `source` nor `data`), and clarify that `--content-safety-off` does not affect image extraction

## How to reproduce
Run `opendataloader-pdf` on a PDF containing images, observe some images have bbox but no `source` in JSON, check logs for the new warnings.

## How to test
```bash
mvn test -Dtest=ImagesUtilsTest#testWriteImageLogsWarningWhenRendererIsUnavailable -pl java/opendataloader-pdf-core
```

## Breaking changes
None — existing behavior preserved; only adds log warnings.

### CODE REVIEW

File: java/opendataloader-pdf-core/src/test/java/org/opendataloader/pdf/utils/ImagesUtilsTest.java [hunk 1]
```diff
@@ -17,11 +18,55 @@
...
+        List<LogRecord> records = new ArrayList<>();
+        Handler handler = new Handler() {
+            public void publish(LogRecord r) { records.add(r); }
+            public void flush() {}
+            public void close() throws SecurityException {}
+        };
+        Logger logger = Logger.getLogger(ImagesUtils.class.getCanonicalName());
+        logger.addHandler(handler);
+        logger.setLevel(Level.WARNING);
+        try {
```
[the remark this answers is not in the document]
Comment (bundolee) [reply java/opendataloader-pdf-core/src/test/java/org/opendataloader/pdf/utils/ImagesUtilsTest.java:52]: Fixed. The original logger level is now captured before mutation and restored in the finally block. Also added try-with-resources for Files.walk() in the same cleanup block.

File: java/opendataloader-pdf-core/src/test/java/org/opendataloader/pdf/utils/ImagesUtilsTest.java [hunk 2]
```diff
@@ -17,11 +18,55 @@
...
+            StaticLayoutContainers.setImagesDirectory(tempDir.toString());
+            // Use an invalid PDF path so contrastRatioConsumer will be null
+            ImagesUtils imagesUtils = new ImagesUtils();
+            ImageChunk chunk = new ImageChunk(new BoundingBox(0));
+            imagesUtils.writeImage(chunk, "/nonexistent/path/file.pdf", "");
+            // Verify ImagesUtils logs a warning when the image cannot be rendered
+            boolean hasImageRenderWarning = records.stream()
+                    .anyMatch(r -> r.getLevel() == Level.WARNING
+                            && r.getLoggerName().equals(ImagesUtils.class.getCanonicalName())
+                            && r.getMessage().toLowerCase().contains("image"));
```
[the remark this answers is not in the document]
Comment (bundolee) [reply java/opendataloader-pdf-core/src/test/java/org/opendataloader/pdf/utils/ImagesUtilsTest.java]: Fixed. The assertion now checks for "PDF renderer unavailable" — the specific substring of the new warning path triggered when contrastRatioConsumer is null. The test is also renamed to testWriteImageLogsWarningWhenRendererIsUnavailable to make the scenario clear.

File: java/opendataloader-pdf-core/src/test/java/org/opendataloader/pdf/utils/ImagesUtilsTest.java [hunk 3]
```diff
@@ -17,11 +18,55 @@
...
+            assertTrue(hasImageRenderWarning, "Should log a warning when image rendering fails (source key will be absent)");
+        } finally {
+            logger.removeHandler(handler);
+            Files.walk(tempDir).sorted((a, b) -> b.compareTo(a)).forEach(p -> {
+                try { Files.deleteIfExists(p); } catch (IOException e) { /* ignore */ }
+            });
```
[the remark this answers is not in the document]
Comment (bundolee) [reply java/opendataloader-pdf-core/src/test/java/org/opendataloader/pdf/utils/ImagesUtilsTest.java]: Fixed. Files.walk() is now wrapped in try-with-resources alongside the logger level restore.

File: content/docs/json-schema.mdx [hunk 4]
```diff
@@ -111,6 +111,16 @@ List items include text properties plus:
...
 
+### When `source` is absent
+
+The `source` field is present only when the image is successfully extracted to a file. It will be absent when:
+
+- **`--image-output embedded`**: The `data` field (Base64) is used instead of `source`.
+- **`--image-output off`**: Image extraction is disabled entirely.
+- **Image rendering failed**: The image is detected in the PDF (bbox is present) but could not be rendered. This can happen with certain PDF image encodings or when the page cannot be rendered. A `WARNING` is logged in this case — check the tool output for `"Image could not be rendered and will be skipped"`.
+
+> **Note:** The `--content-safety-off` option controls text filtering only and has no effect on image extraction or the presence of `source`.
```
[the remark this answers is not in the document]
Comment (bundolee) [reply content/docs/json-schema.mdx]: Verified against Base64ImageUtils.toDataUri() — correct on both counts. Images are written to disk first even in embedded mode (ImagesUtils.write() runs unconditionally when image-output is not off), and toDataUri() returns null for files >10 MB or on IO error, leaving neither source nor data in the output. The inaccurate intro sentence and the incomplete embedded-mode description are now fixed.

**Comment by bundolee** [comment]
All four items from the review addressed in 45e15bf:

**String concatenation in log calls** — switched all three log calls in `createImageFile()` (both new warnings + the pre-existing IOException catch) to parameterized form.

**Two distinct failure cases** — split into separate messages: `contrastRatioConsumer == null` now logs `"Image skipped (PDF renderer unavailable...)"` and `getPageSubImage() == null` logs `"Image could not be rendered..."`. This makes it clear whether all images in the document are affected (renderer initialization failure) or only specific ones (per-image render failure).

**Redundant clearContainers()** — removed. @AfterEach handles it.

**Loose assertion** — test renamed to `testWriteImageLogsWarningWhenRendererIsUnavailable` and assertion checks `"PDF renderer unavailable"` specifically, matching only the new code path exercised by the test scenario.

**Comment by bundolee** [comment]
**PR description test name** — fixed.

**Missing test for `getPageSubImage() == null` path** — the gap is real and worth explaining.

`contrastRatioConsumer` is a private field and `ContrastRatioConsumer` comes from the verapdf library (not easily stubbable without casting or reflection). Testing this path without Mockito means either:

1. **Add Mockito** — clean but introduces a new test dependency this codebase does not currently use.
2. **Make the field package-private** — allows same-package test access; minimal production change, enables a proper test.
3. **Use a real PDF with an out-of-bounds bbox** — fragile, and whether `getPageSubImage()` actually returns null for a specific bbox is implementation-dependent (the existing test shows `BoundingBox(0)` on `lorem.pdf` succeeds and writes a file).

The test that currently exists covers the case that produces the specific `"PDF renderer unavailable"` message — the renderer initialization failure. That is the failure mode most users encounter when a PDF cannot be opened for rendering (and it is the scenario the issue reporter most likely hit).

The `getPageSubImage() == null` path (renderer initialized, specific image unrenderable) requires one of the three options above. Happy to add Mockito or make the field package-private if the team considers the second warning path important enough to pin. Otherwise, both paths are covered by the implementation with distinct, searchable warning strings.

**Comment by bundolee** [comment]
**Review resolution summary**

All review feedback has been addressed. Changes since the original commit:

| Reviewer | Item | Resolution |
|----------|------|------------|
| claude | String concat in log calls | Fixed — all three calls use parameterized form |
| claude | Single message conflates two failure modes | Fixed — split into "PDF renderer unavailable" (consumer==null) and "Image could not be rendered" (getPageSubImage()==null) |
| claude | Redundant clearContainers() in test | Fixed — removed; @AfterEach handles it |
| claude | Loose assertion on "image" | Fixed — assertion checks "PDF renderer unavailable" specifically |
| Copilot | Logger level not restored | Fixed — original level saved and restored in finally |
| Copilot | Files.walk() stream not closed | Fixed — wrapped in try-with-resources |
| Copilot | Docs: embedded mode description inaccurate | Fixed — clarified that embedded mode still omits source intentionally and that Base64 failure leaves neither field |
| claude (round 2) | PR description test name mismatch | Fixed |
| claude (round 2) | Missing test for getPageSubImage()==null path | Acknowledged gap; ContrastRatioConsumer is a verapdf library class not easily stubbable without Mockito (not in this project's POM). Current test covers the renderer-unavailable path with a specific assertion. The second path requires either adding Mockito or making contrastRatioConsumer package-private — open to either if prioritized. |

PR is ready for merge.
